src/python/pants/jvm/resolve/lockfile_metadata.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Iterable, cast
+
+from pants.core.util_rules.lockfile_metadata import (
+    LockfileMetadata,
+    LockfileMetadataValidation,
+    LockfileScope,
+    _get_metadata,
+    lockfile_metadata_registrar,
+)
+from pants.jvm.resolve.common import ArtifactRequirement
+from pants.util.ordered_set import FrozenOrderedSet
+
+logger = logging.getLogger(__name__)
+
+_jvm_lockfile_metadata = lockfile_metadata_registrar(LockfileScope.JVM)
+
+
+class InvalidJVMLockfileReason(Enum):
+    REQUIREMENTS_MISMATCH = "requirements_mismatch"
+
+
+@dataclass(frozen=True)
+class JVMLockfileMetadata(LockfileMetadata):
+
+    scope = LockfileScope.JVM
+
+    @staticmethod
+    def new(
+        requirements: Iterable[ArtifactRequirement],
+    ) -> LockfileMetadata:
+        """Call the most recent version of the `LockfileMetadata` class to construct a concrete
+        instance.
+
+        This static method should be used in place of the `LockfileMetadata` constructor. This gives
+        calling sites a predictable method to call to construct a new `LockfileMetadata` for
+        writing, while still allowing us to support _reading_ older, deprecated metadata versions.
+        """
+
+        return JVMLockfileMetadataV1.from_artifact_requirements(requirements)
+
+    @classmethod
+    def from_lockfile(
+        cls, lockfile: bytes, lockfile_path: str | None = None, resolve_name: str | None = None
+    ) -> JVMLockfileMetadataV1:
+        return cast(
+            JVMLockfileMetadataV1,
+            LockfileMetadata.from_lockfile_for_scope(
+                LockfileScope.JVM, lockfile, lockfile_path, resolve_name
+            ),
+        )
+
+    def _header_dict(self) -> dict[Any, Any]:
+        """Produce a dictionary to be serialized into the lockfile header.
+
+        Subclasses should call `super` and update the resulting dictionary.
+        """
+
+        d = super()._header_dict()
+        return d
```

Comment:
Why override this?

Reply:
You have to override it in subclasses, and you need to keep `super` in place. This override here is defensive work to make sure that people who eventually add things into `_header_dict` in this class don't do it the wrong way.

Reply:
In general, I find that it is preferable in OO to avoid APIs that require overriding concrete methods, and to stick to completely abstract (or empty) methods or fields. But not a blocker.